### PR TITLE
storage: support order keys in upsert state api

### PR DESCRIPTION
**Pull request by guswynn**
This is broken out of: https://github.com/MaterializeInc/materialize/pull/24663

Fundamentally, fixing https://github.com/MaterializeInc/materialize/issues/23998 requires that the upsert state backends are able to _ingest data that arrives before a progress update_. This requires we store and _order key_ alongside that data, so that we can order values for the same key that occurred at the same _source time_, but within the same mz timestamp. This includes tombstones for keys that were deleted. This pr adds this functionality to the upsert state backend, but does not make use of it yet.

Tombstones unfortunately take up permanent space in the state. There is no easy way around this (other than doing a full range scan of the state on `Progress` updates), and our choice to avoid doing so will be discussed in the next pr (24663)

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by moulimukherjee — APPROVED** [review]
lgtm!

File: src/storage/src/render/upsert/types.rs [hunk 1]
```diff
@@ -381,8 +440,10 @@ pub struct PutStats {
     /// The number of puts/deletes processed
     /// Should be equal to number of inserts + updates + deletes
     pub processed_puts: u64,
-    /// The aggregated number of values inserted or deleted into `state`
+    /// The aggregated number of non-tombstone values inserted or deleted into `state`
```
Comment (moulimukherjee) [inline src/storage/src/render/upsert/types.rs:443]:
```suggestion
    /// The aggregated number of non-tombstone values inserted or deleted into `state`.
```

File: src/storage/src/render/upsert/types.rs [hunk 2]
```diff
@@ -381,8 +440,10 @@ pub struct PutStats {
...
     pub values_diff: i64,
+    /// The aggregated number of tombstones inserted or deleted into `state`
```
Comment (moulimukherjee) [inline src/storage/src/render/upsert/types.rs:445]:
```suggestion
    /// The aggregated number of tombstones inserted or deleted into `state`.
```

File: src/storage/src/render/upsert.rs [hunk 3]
```diff
@@ -422,25 +429,49 @@ async fn drain_staged_input<S, G, T, O, E>(
...
+        let existing_order = existing_value.as_ref().map(|cs| cs.order().as_ref());
+
+        // If the value does not exist, or has the default `O` (None::<FromTime>`), we
+        // always sort below the from time of the new value. This ensures that new values
+        // overwrite keys with no values (and no tombstones), and all values that are
+        // recorded in rehydration (which also use the default `O`)
+        if existing_order >= Some(Some(&from_time.0)) {
+            command_state.remove();
+            continue;
+        }
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:455]: I can't parse this comment, where do we check for the default value? It also says "if X, we always sort".

File: src/storage/src/render/upsert.rs [hunk 4]
```diff
@@ -422,25 +429,49 @@ async fn drain_staged_input<S, G, T, O, E>(
...
+
         match value {
             Some(value) => {
-                if let Some(old_value) = command_state.value.replace(value.clone().into()) {
-                    output_updates.push((old_value.to_decoded(), ts.clone(), -1));
+                if let Some(old_value) = existing_value
+                    .replace(StateValue::value(value.clone(), Some(from_time.0.clone())))
+                {
+                    // Similarly, we never record a tombstone value with an order key in this commit.
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:462]: Similarly to what? Why don't we have to record a tombstone?

**Review by petrosagg — COMMENTED** [review]
I'm finding it difficult to follow what is going on here. I think the upsert modules are lacking some high level description of what is going on with the various operators like the mysql and postgres sources have. Can we have some prose describing upsert, including this new tombstone handling logic?

File: src/storage/src/render/upsert.rs (hunk 3 above)
Comment (guswynn) [reply src/storage/src/render/upsert.rs:455]: will adjust!

**Comment by guswynn** [comment]
> Can we have some prose describing upsert, including this new tombstone handling logic?


yes!

**Comment by guswynn** [comment]
@petrosagg fixed your questions and fixed up the module docs!

**Review by petrosagg — APPROVED** [review]
LGTM! Thanks for the extra documentation

**Pull request merged**